proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,274 @@
+be distributed out over multiple blocks.
+
+In the new stake rewards distribution approach, we will separate the
+computation of rewards from the actual distribution of rewards at the epoch
+boundary by dividing the process into two distinct phases:
+
+   1. rewards calculation phase - during which the epoch rewards for all
+      activate stake accounts are computed and distribution chunks are scheduled.
+
+   1. rewards distribution phase - during which the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+To help maintain the total capital balance and track/verify the reward
+distribution during rewarding phases, a new sysvar account, `EpochRewards`, is
+proposed. The `EpochRewards` sysvar holds the balance of the rewards that are
+pending for distribution.
+
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks for distribution in the next phase.
+
+Currently, on Solana Mainnet Beta with ~550K active stake accounts, epoch reward
+calculation takes around 10 seconds on average. This will make it impossible to
+perform rewards computation synchronous within one block.
+
+However, there are quite a few promising optimizations that can cut down the
+reward computation time. An experiment for reward calculation optimization
+(https://github.com/solana-labs/solana/pull/31193) showed that we can cut the
+reward calculation time plus vote reward distribution time to around 1s. This
+makes synchronous reward computation and asynchronous reward distribution a
+feasible approach. We also believe that there is room for more optimization to
+further cut down the above timing.
+
+Therefore, the following design is based on the above optimization. The reward
+calculation will be performed at the first block of the epoch, block height `X`.
+Once the full rewards are calculated, the rewards will be partitioned into
+distribution chunks stored in the bank, which will then be distributed during
+the `reward distribution` phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows.
+
+To minimize the impact on block processing time during the reward distribution
+phase, a target of 4,096 stake rewards will be distributed per block. The total
+number of blocks `M` needed to distributed rewards is given by the following
+formula to round up to the nearest integer without using floating point value
+arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+To safeguard against the number of stake accounts growing dramatically and
+overflowing the number of blocks in an epoch, the number of blocks is capped at
+10% of the number of block in an epoch (currently 432,000).
```

Comment:
we make no provision for how this overflow case is actually handled. iirc the original proposal was to "let the last partition take a long time," but the partitioning algorithm doesn't accommodate.

Reply:
What's written here does describe the current implementation. "Overflow" is actually handled by letting the number of accounts in a partition exceed 4096 by clamping the number of partitions to `slots_per_epoch/10`:
https://github.com/solana-labs/solana/blob/8ad125d0c0688aaf2b62bb95b535ff988ed7f9ac/runtime/src/bank.rs#L1162-L1165

So each slot will go a little slower, rather than distribution extending for more slots. This is probably okay for now, but we probably ought to consider making `MAX_FACTOR_OF_REWARD_BLOCKS_IN_EPOCH` tunable at some point.